### PR TITLE
Update DMS

### PR DESCRIPTION
**Pull request by junha1**

### CODE REVIEW

File: network/src/dms.rs [hunk 1]
```diff
@@ -331,25 +316,119 @@ impl<N: GossipNetwork, S: Storage> DistributedMessageSet<N, S> {
...
+            }
+        }
+        Ok(())
+    }
+
     /// Reads the messages from the storage.
     pub async fn read_messages(&self) -> Result<Vec<Message>, Error> {
-        let result = read_messages(&*self.storage.read().await).await?;
-        Ok(result)
+        let files = self.storage.read().await.list_files().await?;
```
Comment (lemonholic) [inline network/src/dms.rs:394]: Why don't we filter `STATE_FILE_PATH ` here?

File: network/src/dms.rs [hunk 2]
```diff
@@ -331,25 +316,119 @@ impl<N: GossipNetwork, S: Storage> DistributedMessageSet<N, S> {
...
+                .await
+                .map(|r| (f, r))
+        });
+        let data = future::join_all(tasks)
+            .await
+            .into_iter()
+            .collect::<Result<Vec<_>, _>>()?;
+        let messages = data
+            .into_iter()
+            .filter_map(|(f, r)| if f == STATE_FILE_PATH { None } else { Some(r) })
```
Comment (lemonholic) [inline network/src/dms.rs:409]: And not here.

File: network/src/dms.rs [hunk 3]
```diff
@@ -359,109 +438,170 @@ impl<N: GossipNetwork, S: Storage> DistributedMessageSet<N, S> {
...
-                        Result::<(), Error>::Ok(())
-                    }
-                });
-                join_all(tasks).await;
-                tokio::time::sleep(interval).await;
+        loop {
+            if let Err(e) = Self::fetch(
+                &mut *this.write().await,
+                &network_config,
+                &peers.read().await,
```
Comment (lemonholic) [inline network/src/dms.rs:497]: Don't know how smart Rust is, but does it automatically drops the locks when the fetch is done, in the loop?

File: network/src/dms.rs [hunk 4]
```diff
@@ -359,109 +438,170 @@ impl<N: GossipNetwork, S: Storage> DistributedMessageSet<N, S> {
...
+        let (switch_send, switch_recv) = tokio::sync::oneshot::channel();
+        let this_ = Arc::clone(&this);
+        Ok(Serve::new(
+            tokio::spawn(async move {
+                let mut tasks =
+                    future::try_join4(rpc_task, fetch_task, broadcast_task, gossip_task).fuse();
+                let mut switch_recv = switch_recv.fuse();
+                select! {
+                    x = tasks => {
+                        let _ = switch_recv.await;
```
Comment (lemonholic) [inline network/src/dms.rs:590]: Why do you await the termination signal here?

File: network/src/dms.rs (hunk 3 above)
Comment (junha1) [reply network/src/dms.rs:497]: yes.

File: network/src/dms.rs (hunk 1 above)
Comment (junha1) [reply network/src/dms.rs:394]: fixed.

File: network/src/dms.rs (hunk 2 above)
Comment (junha1) [reply network/src/dms.rs:409]: fixed.

File: network/src/dms.rs (hunk 4 above)
Comment (junha1) [reply network/src/dms.rs:590]: I added comment

File: network/src/dms.rs (hunk 3 above)
Comment (lemonholic) [reply network/src/dms.rs:492]: okok

File: network/src/dms.rs (hunk 4 above)
Comment (lemonholic) [reply network/src/dms.rs:588]: Got it.

**Review by lemonholic — APPROVED** [review]
looking good

**Pull request merged**